server/src/main/java/org/opensearch/index/store/remote/utils/cache/LRUCache.java

```diff
@@ -0,0 +1,423 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store.remote.utils.cache;
+
+import org.opensearch.common.cache.RemovalListener;
+import org.opensearch.common.cache.RemovalNotification;
+import org.opensearch.common.cache.RemovalReason;
+import org.opensearch.common.cache.Weigher;
+import org.opensearch.index.store.remote.utils.cache.stats.CacheStats;
+import org.opensearch.index.store.remote.utils.cache.stats.DefaultStatsCounter;
+import org.opensearch.index.store.remote.utils.cache.stats.StatsCounter;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.BiFunction;
+
+/**
+ * LRU implementation of {@link RefCountedCache}. As long as {@link Node#refCount} greater than 0 then node is not eligible for eviction.
```

Comment:
Can we list out the requirements this cache meets that you detailed [in this comment](https://github.com/opensearch-project/OpenSearch/issues/4964#issuecomment-1421689586), and maybe drop a link to #6225 so that it is clear to any reader of the code why this exists?

Reply:
totally make sense.